boards/boardsource-blok/examples/blok_usb_serial.rs

```diff
@@ -0,0 +1,160 @@
+static mut USB_SERIAL: Option<SerialPort<hal::usb::UsbBus>> = None;
+
+#[entry]
+fn main() -> ! {
+    let mut pac = pac::Peripherals::take().unwrap();
+
+    let mut watchdog = Watchdog::new(pac.WATCHDOG);
+
+    let clocks = init_clocks_and_plls(
+        XOSC_CRYSTAL_FREQ,
+        pac.XOSC,
+        pac.CLOCKS,
+        pac.PLL_SYS,
+        pac.PLL_USB,
+        &mut pac.RESETS,
+        &mut watchdog,
+    )
+    .ok()
+    .unwrap();
+
+    let sio = Sio::new(pac.SIO);
+    let _pins = Pins::new(
+        pac.IO_BANK0,
+        pac.PADS_BANK0,
+        sio.gpio_bank0,
+        &mut pac.RESETS,
+    );
+
+    let _timer = Timer::new(pac.TIMER, &mut pac.RESETS);
+
+    let usb_bus = UsbBusAllocator::new(hal::usb::UsbBus::new(
+        pac.USBCTRL_REGS,
+        pac.USBCTRL_DPRAM,
+        clocks.usb_clock,
+        true,
+        &mut pac.RESETS,
+    ));
+    unsafe {
+        USB_BUS = Some(usb_bus);
+    }
+
+    let bus_ref = unsafe { USB_BUS.as_ref().unwrap() };
+
+    let serial = SerialPort::new(bus_ref);
+    unsafe {
+        USB_SERIAL = Some(serial);
+    }
+
+    let usb_device = UsbDeviceBuilder::new(bus_ref, UsbVidPid(0x1209, 0x0001))
+        .product("serial port")
+        .device_class(2) // from: https://www.usb.org/defined-class-codes
+        .build();
+    unsafe {
+        USB_DEVICE = Some(usb_device);
+    }
+
+    unsafe {
+        pac::NVIC::unmask(hal::pac::Interrupt::USBCTRL_IRQ);
+    }
+
+    let core = pac::CorePeripherals::take().unwrap();
+    let mut delay = cortex_m::delay::Delay::new(core.SYST, clocks.system_clock.freq().to_Hz());
+
+    let mut i: u8 = 0;
+
+    loop {
+        delay.delay_ms(2_000);
+
+        let mut text: String<20> = String::new();
+        writeln!(&mut text, "loop number: {}\r\n", i).unwrap();
+
+        write_serial(text.as_bytes());
+
+        i += 1;
+        if i >= 10 {
+            hal::rom_data::reset_to_usb_boot(0, 0);
+        }
+    }
+}
+
+/// Writes to the serial port.
+///
+/// We do this with interrupts disabled, to avoid a race hazard with the USB IRQ.
+fn write_serial(byte_array: &[u8]) {
+    let _ = critical_section::with(|_| unsafe {
+        USB_SERIAL.as_mut().map(|serial| serial.write(byte_array))
+    })
+    .unwrap();
+}
+
+/// This function is called whenever the USB Hardware generates
+/// an Interrupt Request
+#[allow(non_snake_case)]
+#[interrupt]
+unsafe fn USBCTRL_IRQ() {
+    let usb_device = USB_DEVICE.as_mut().unwrap();
+    let serial = USB_SERIAL.as_mut().unwrap();
+
+    if usb_device.poll(&mut [serial]) {
+        let mut buf = [0u8; 64];
+
+        match serial.read(&mut buf) {
+            Err(_e) => {}
+            Ok(_count) => {
+                // gets the first 4 bytes of buf
+                let mut read_text = [0u8; 4];
+                read_text
+                    .iter_mut()
+                    .enumerate()
+                    .for_each(|(i, e)| *e = buf[i]);
+
+                if &read_text == b"stop" {
```

Comment:
if bytes are received 1 by 1, this will not work as expected.

This behaviour is dependant on many factors such as:
- Host platform (windows/linux/macos etc)
- serial port tool (Hercules, putty, minicom, screen etc)

Also, although I think the reading of the serial port needs to be done in the interrupt in order to garanty no data loss may occur, the processing of that data would be better off in the main loop rather than the interrupt.
In this specific case, it does not really mater, but it's always good to demonstrate good practices in examples not to confuse less seasoned developers.

Reply:
To avoid the issue with bytes being received 1 by 1 should they be stored between interrupts or is there a more optimal way?

There is an interrupt serial example for the pico and the processing of the data was also kept in the interrupt handler, should this then be changed as well?

Reply:
You would either manage a state machine that steps forward when the next character of `stop` is received but that might be a bit too heavy for the purpose of that example :D 

A static buffer should be enough. Whether the `static` buffer is owned locally to the interrupt or global, I don't really mind.